Review request -- Mathieu fcns

brorson opened the pull request:
This is a request to try the new Mathieu fcn impl.   This new fcn impl has  been discussed in https://github.com/scipy/xsf/issues/47 .  I have updated the docstrings and other comment in this checkin, but the functionality has otherwise remained the same as that shown in the bug report.

I have run these fcns against the Scipy test suite using "spin test" and the test suite runs to completion.  The Mathieu fcns fail, but that is expected since this is a new impl and the input arg is in radians, not degrees.

Please review and comment.  My goal here is to request a review and find out what I need to change.

Stuart

lucascolley commented:
I'm afraid I can't help with C++ expertise, but I wonder whether you could open a draft PR to SciPy which shows the tests running (albeit with failures)? No rush, just thought it might be worth it whilst you are waiting for review here and have some momentum. Let me know if you would like help with making the necessary changes to the SciPy submodule etc.

brorson commented:
Thank you for your suggestion, @lucascolley.  I will do that.

Meanwhile, I see the CI thing is running on my PR and I have build failures.  I fixed one -- I was using the suffix 'd' when declaring double literals which turns out to be non-standard.  I fixed that one and pushed to my forked repo.  Not sure if CI runs continuously but I will continue fixing any bugs it finds.

Stuart

brorson commented:
OK, I just submitted a PR on the main Scipy branch.  The pull request includes:

1. My changes to the meson.build file which are required to successfully build the Mathieu stuff.  The change is imply to add the Lapack dependencies to ufuncs and other build targets per @steppi 's suggestion.
2. I also mentioned the failing tests.  The Mathieu tests now fail since the implementation is different and returns different results from the old impl.  (The old tests didn't check correctness; they just checked that the output of the fcns didn't change.)  The required action item here is to bypass the tests for the time being so the impls can be integrated into the codebase.  I also offer to provide some Python-based tests written by my students which can be used in the future.  Maybe it would make sense to have a Scipy branch we can work in and not bother the trunk?

I also submitted a new PR on the xsf  repo to pick up a bugfix I implemented after looking at some of my last CI test failures:  I had suffixed some double literals using 'd', like this: 2.0d.  This is apparently not standard C++ so made some of the compilers barf.  My mistake -- I fixed it and issued a new PR to you.

 Sorry for the multiple PRs -- I'm new to the Scipy dev chain and so will fumble around a little bit as I try to get this change into the codebase.

Stuart

steppi commented:
Thanks @brorson. I plan to spend some time in the next few weeks to improve testing, and we can add some test cases based on references from https://github.com/mathieuandspheroidalwavefunctions computed with quad precision. In the meantime, your `main.c` should be moved into the `tests` folder and turned into a proper [Catch2](https://github.com/catchorg/Catch2) test file named something like `xsf/tests/miscellaneous_tests/test_mathieu_funcs.cpp` that will get picked up by the test harness. You don't have to use parquet tables, you can just hardcode the golden values into the test files. Let me know if you need any help with this. If you do this, I can set the tolerances in the relevant tests in `xsf/tests/scipy_special_tests` to infinity until reference values are recomputed.

rgommers commented on include/xsf/mathieu/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Is there a reason why you're adding a new license file, rather than contributing this under `xsf`'s license? The latter is normally expected and would be a lot easier (license files propagate, even beyond this repo).

steppi replied:
I was going to comment that if @brorson has specific reasons to prefer an MIT license for this code, an entry will need to be added to https://github.com/scipy/xsf/blob/main/LICENSES_bundled.txt. It would of course be simpler to contribute this under `xsf`'s license though

brorson commented:
Thank you, @steppi.  Regarding the tests in main.c, sure I can move them to a file called "test_mathieu_fcns.cpp" and move them to xsf/tests/miscellaneous_tests/.  Give me a day or two.

Once I do that I will check the mods into my GitHub repo and issue another PR so you know they're there.

jorenham commented on include/xsf/mathieu/mathieu_fcns.h:
```diff
@@ -0,0 +1,1600 @@
+ * This is part of the Mathieu function suite -- a reimplementation
+ * of the Mathieu functions for Scipy.  This file holds the function
+ * implementations themselves.  The prototype was written in Matlab
+ * and validated.  This is a translation from Matlab to C.
```
```suggestion
 * and validated.  This is a translation from Matlab to C++.
```

😉

jorenham commented on include/xsf/mathieu/mathieu_fcns.h:
```diff
@@ -0,0 +1,1600 @@
+namespace mathieu {
+
+    //==================================================================
+    int mathieu_ce(int m, double q, double v, double *ce, double *ced) {
```
xsf uses pass-by-reference in case of multiple scalar outputs

brorson commented:
Thank you for the suggestions, @jorenham .  I will go down the list and fix up my code based on your comments.

Regarding code duplication, yes I thought of using macros for the repeated stuff but since the repeated stuff is multi-line I figured any macro would be multi-line and hard to manage.  However, I'll think about how to use macros again based on your comments, and may also see if I can stick some of the stuff into a separate inline fcn.  I will want to comment heavily any macro or fcn to make clear what the calc is trying to do.

Regarding malloc() & free(), yeah, I am more of a C programmer than a C++ programmer.  Does Scipy allow use of smart pointers?  I see auto_ptr has been removed from C++ ... is there a preferred alternative?

(I will browse the codebase to see what is used, but it's always easier to ask ...)

Regarding failing loudly, do you mean throw an exception?  I may be wrong, but since the xsf stuff supplies return codes I thought that was the preferred way to deal with errors.  Please correct me if wrong.

Thx,
Stuart

brorson commented:
Regarding code structure, mathieu.h living in the xsf directory provides the interface to the rest of Scipy.  It is just a wrapper around my impls.  My impls live in the mathieu subdirectory.  Passing error codes within my stuff I have handled by just using integer flags, but I should fix that to conform to Scipy's error codes.  Error codes passed back to Scipy from the interface fcns in mathieu.h are currently done using Scipy/xsf's predefined codes in error.h.  I made a guess about what the different errors in the enum mean ... I hope I was close enough to be correct.

brorson commented:
Regarding status of this project, I am currently bogged down with teaching and other responsibilities but hope to get back to this in a couple of weeks.  Besides some clean-up per your suggestions (and my own list of things to clean up) I need to follow @rgommers suggestion about how to create a pull request which makes changes to two repos -- the main Scipy repo (where the Meson build config lives) and the xsf repo.  I need to declare some dependencies in the build config stuff.  That's waiting for me to do.

brorson commented on include/xsf/mathieu.h:
```diff
@@ -1,228 +1,352 @@
+ * @return      Mathieu eigenvalue a.
+ */
+template <typename T>
+T cem_cva(T m, T q) {
```
You are right, but this is how it was done in the old code, so I am reluctant to change it.  This fcn is just a wrapper -- an adaption layer to the rest of Scipy.

brorson commented on include/xsf/mathieu/besseljyd.h:
```diff
@@ -0,0 +1,84 @@
+    //==================================================================
+    double besselj(int k, double z) {
+        // This is just a thin wrapper around the Bessel impl in the
+        // std library.
```
Fixed.

brorson commented on include/xsf/mathieu/make_matrix.h:
```diff
@@ -0,0 +1,198 @@
+ *
+ */
+
+#define SQRT2 1.414213562373095
```
Fixed.

brorson commented on include/xsf/mathieu/make_matrix.h:
```diff
@@ -0,0 +1,198 @@
+    A = recurrence matrix (must be calloc'ed in caller).
+    Return:
+    return code = 0 if OK.
+    -------------------------------------------------*/
```
True, but my screen-size is limited since I work on a laptop and not a nice 4-screen work area.  I prefer to keep the comment as is, if that's OK.

brorson commented on include/xsf/mathieu/make_matrix.h:
```diff
@@ -0,0 +1,198 @@
+    return code = 0 if OK.
+    -------------------------------------------------*/
+    int make_matrix_ee(int N, double q, double *A) {
+        int j;
```
Not sure where to move it -- it gets used in a couple of other fcns defined in other files ...  That's the point of the make_matrix file -- it holds some utility fcns used in other places.  Perhaps I am missing your point (very likely) ? ....

brorson commented on include/xsf/mathieu/make_matrix.h:
```diff
@@ -0,0 +1,198 @@
+        i = MATRIX_IDX(N, N - 1, N - 2);
+        A[i] = q;
+        i = MATRIX_IDX(N, N - 1, N - 1);
+        A[i] = (2.0 * (N - 1)) * (2.0 * (N - 1));
```
True, but I find this form easier to parse.  Expanding it all out would mean I'd need to reverse-engineer an a*x^2 + b*x + c quadratic to understand it's just a simple square.

brorson commented on include/xsf/mathieu/mathieu.h:
```diff
@@ -0,0 +1,352 @@
+template <typename T>
+T cem_cva(T m, T q) {
+    // This returns the even Mathieu characteristic value (eigenvalue) a.
+
```
Yeah, you are right.  One thing I hate is that the eigenvalues appearing in the Mathieu ODE are traditionally called its "characteristic values".  But they're really eigenvalues if you view the equation as an operator equation.  Therefore, I used both terms in case somebody (like some students) aren't aware of both terms.  It is redundant but it helps clarify what is happening so I am inclined to leave it in.

brorson commented on include/xsf/mathieu/mathieu.h:
```diff
@@ -0,0 +1,352 @@
+ * @return      Mathieu eigenvalue a.
+ */
+template <typename T>
+T cem_cva(T m, T q) {
```
Fixed.

brorson commented on include/xsf/mathieu/mathieu_coeffs.h:
```diff
@@ -0,0 +1,295 @@
+
+        // Bail out if m is not even.
+        if (m % 2 != 0)
+            return -1;
```
Good point.  I fixed this by using the xsf stuff for returning specific error types.
